### PR TITLE
2021 04 peptide search form and peptide search and id mapping submission

### PR DESCRIPTION
**Pull request by aurel-l**
## Purpose
Create a Peptide Search form and implement submission logic. [TRM-25699](https://www.ebi.ac.uk/panda/jira/browse/TRM-25699) and [TRM-25640](https://www.ebi.ac.uk/panda/jira/browse/TRM-25640)
Implement ID Mapping submission logic. [TRM-25700](https://www.ebi.ac.uk/panda/jira/browse/TRM-25700)

## Approach
Same as purpose

## Testing
No specific test while specs are fluid

## Checklist
- [ ] My PR is scoped properly, and “does one thing only”
- [x] I have reviewed my own code
- [x] I have checked that linting checks pass and type safety is respected
- [ ] I have checked that tests pass and coverage has at least improved, and if not explained the reasons why
  - Lots of new logic in job middleware, difficult to unit-test

Current issues and doubts:
 - [ ] unable to use the endpoint, so fall back to current UniProt website as make-do endpoint (will get issue fetching results, seems like they stay forever running unless we do a browser browse of the URL)
 - [ ] As per Jira, supposed to use the existing sequence field and validation, but peptide search doesn't currently support FASTA, and does support smaller sequence -> What we currently have doesn't fit.
 - [ ] We have a component in BLAST and Align forms to fetch and pre-fill sequences from existing entries. Should we have it here too? Might be a tool for the user to fetch the sequence then trim it to the wanted subsequence, but maybe user will think this is the canonical/only way to do peptide searches

Note: To submit an ID Mapping job on this PR, submit any Peptide Search through the form and it will create a predefined ID Mapping job alongisde it

### CODE REVIEW

File: src/tools/id-mapping/types/idMappingServerParameters.ts [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+export type ServerParameters = {
+  from: From;
+  to: To;
+  ids: IDs;
+};
+
+// same, because no email was provided before
+export type PublicServerParameters = ServerParameters;
+
+export type IDMappingTarget =
```
Comment (xwatkins) [inline src/tools/id-mapping/types/idMappingServerParameters.ts:19]: I guess that list will grow with Dan's PR?

File: src/tools/peptide-search/components/__tests__/PeptideSearchForm.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+import initialState from '../../../../app/state/rootInitialState';
+
+import { mockSuggesterApi } from '../../../../query-builder/components/__tests__/__mocks__/autocompleteWrapperData';
+
+const mock = new MockAdapter(axios);
+mock.onGet().reply(200, mockSuggesterApi.response);
+
+const aaSequence = 'ABCDEFGHIJKLMNOPQRST';
+const ntSequence = 'ATCGAGCGATAGCGAGGGAC';
+
```
Comment (xwatkins) [inline src/tools/peptide-search/components/__tests__/PeptideSearchForm.spec.tsx:19]: Should we try with an invalid sequence too?

File: src/tools/peptide-search/config/PeptideSearchFormData.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  spOnly = 'Search UniProt Reviewed (Swiss-Prot) only',
+  name = 'Name',
+}
+
+export type PeptideSearchFormValues = Record<
+  PeptideSearchFields,
+  Readonly<PeptideSearchFormValue>
+>;
+
+const formData: Readonly<PeptideSearchFormValues> = Object.freeze({
```
Comment (xwatkins) [inline src/tools/peptide-search/config/PeptideSearchFormData.ts:33]: Can you remind me why the `Object.freeze`?

File: src/tools/state/getCheckJobStatus.ts [hunk 4]
```diff
@@ -13,6 +13,7 @@ import { RunningJob, FinishedJob } from '../types/toolsJob';
 import { Status } from '../types/toolsStatuses';
 import { BlastResults } from '../blast/types/blastResults';
 import { JobTypes } from '../types/toolsJobTypes';
+import { IDMappingTarget } from '../id-mapping/types/idMappingServerParameters';
```
Comment (xwatkins) [inline src/tools/state/getCheckJobStatus.ts:16]: Not needed (not sure why eslint isn't complaining though, is this file not checked?)

**Review by xwatkins — CHANGES_REQUESTED** [review]

File: src/tools/peptide-search/config/PeptideSearchFormData.ts (hunk 3 above)
Comment (aurel-l) [reply src/tools/peptide-search/config/PeptideSearchFormData.ts:33]: To make sure to not mutate the object while processing it in the form, this is due to the complexity of the form description object and the difficulty to track unexpected side effects related to accidentally mutating it

File: src/tools/id-mapping/types/idMappingServerParameters.ts (hunk 1 above)
Comment (aurel-l) [reply src/tools/id-mapping/types/idMappingServerParameters.ts:19]: Open to suggestions for a better variable name

Comment (xwatkins) [reply src/tools/id-mapping/types/idMappingServerParameters.ts:19]: Ah I see - `IDMappingResultsNamespace`? `MappingResultsNamespace`?

File: src/tools/peptide-search/components/__tests__/PeptideSearchForm.spec.tsx (hunk 2 above)
Comment (aurel-l) [reply src/tools/peptide-search/components/__tests__/PeptideSearchForm.spec.tsx:19]: All those tests are skipped now, pending discussion point #2

File: src/tools/state/getCheckJobStatus.ts (hunk 4 above)
Comment (aurel-l) [reply src/tools/state/getCheckJobStatus.ts]: Not sure why that happened either...
Removed now

File: src/tools/id-mapping/types/idMappingServerParameters.ts (hunk 1 above)
Comment (aurel-l) [reply src/tools/id-mapping/types/idMappingServerParameters.ts]: Changed

**Review by xwatkins — APPROVED** [review]

**Pull request merged**